File: packages/organization-graph/src/canonical.ts
import type {
  ActorRef,
  AuthorityRef,
  GraphCompleteness,
  GraphDelta,
  GraphEdge,
  GraphNode,
  GraphSnapshot,
} from './types.js';
import { validateGraphDelta, validateGraphSnapshot } from './validation.js';
import { canonicalJson } from './canonical-json.js';
export { canonicalJson } from './canonical-json.js';

function compare(left: string, right: string): number {
  return left < right ? -1 : left > right ? 1 : 0;
}

function sortedUnique(values: readonly string[]): string[] {
  return [...new Set(values)].sort(compare);
}

function canonicalProperties(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(canonicalProperties);
  if (value !== null && typeof value === 'object') {
    return Object.fromEntries(
      Object.keys(value as Record<string, unknown>)
        .sort(compare)
        .map((key) => [key, canonicalProperties((value as Record<string, unknown>)[key])]),
    );
  }
  return value;
}

function canonicalAuthority(value: AuthorityRef): AuthorityRef {
  return {
    provider: value.provider,
    objectType: value.objectType,
    objectId: value.objectId,
    version: value.version,
    observedAt: value.observedAt,
  };
}

function canonicalOwners(values: readonly ActorRef[]): ActorRef[] {
  const keyed = new Map<string, ActorRef>();
  for (const value of values) {
    const canonical: ActorRef = {
      id: value.id,
      kind: value.kind,
      ...(value.displayName === undefined ? {} : { displayName: value.displayName }),
    };
    keyed.set(canonicalJson(canonical), canonical);
  }
  return [...keyed.values()].sort((left, right) =>
    compare(
      `${left.kind}\u0000${left.id}\u0000${left.displayName ?? ''}`,
      `${right.kind}\u0000${right.id}\u0000${right.displayName ?? ''}`,
    ),
  );
}

function canonicalNode(value: GraphNode): GraphNode {
  return {
    id: value.id,
    type: value.type,
    scenarioDefinitionId: value.scenarioDefinitionId,
    scenarioInstanceId: value.scenarioInstanceId,
    title: value.title,
    ...(value.status === undefined ? {} : { status: value.status }),
    authorityRef: canonicalAuthority(value.authorityRef),
    owners: canonicalOwners(value.owners),
    properties: canonicalProperties(value.properties) as Record<string, unknown>,
    sourceEventIds: sortedUnique(value.sourceEventIds),
    evidenceRefs: sortedUnique(value.evidenceRefs),
    projectorVersion: value.projectorVersion,
    validFrom: value.validFrom,
    ...(value.validTo === undefined ? {} : { validTo: value.validTo }),
  };
}

function canonicalEdge(value: GraphEdge): GraphEdge {
  return {
    id: value.id,
    type: value.type,
    from: value.from,
    to: value.to,
    scenarioInstanceId: value.scenarioInstanceId,
    ...(value.authorityRef === undefined
      ? {}
      : { authorityRef: canonicalAuthority(value.authorityRef) }),
    sourceEventIds: sortedUnique(value.sourceEventIds),
    evidenceRefs: sortedUnique(value.evidenceRefs),
    projectorVersion: value.projectorVersion,
    validFrom: value.validFrom,
    ...(value.validTo === undefined ? {} : { validTo: value.validTo }),
  };
}

function canonicalCompleteness(value: GraphCompleteness): GraphCompleteness {
  return {
    sourcesRequested: sortedUnique(value.sourcesRequested),
    sourcesObserved: sortedUnique(value.sourcesObserved),
    missingSources: sortedUnique(value.missingSources),
    warnings: sortedUnique(value.warnings),
  };
}

export function canonicalizeGraphSnapshot(value: unknown): GraphSnapshot {
  const snapshot = validateGraphSnapshot(value);
  return {
    schema: snapshot.schema,
    cursor: snapshot.cursor,
    scenarioInstanceId: snapshot.scenarioInstanceId,
    generatedAt: snapshot.generatedAt,
    projectorVersion: snapshot.projectorVersion,
    nodes: snapshot.nodes.map(canonicalNode).sort((left, right) => compare(left.id, right.id)),
    edges: snapshot.edges.map(canonicalEdge).sort((left, right) => compare(left.id, right.id)),
    completeness: canonicalCompleteness(snapshot.completeness),
    integrityHash: snapshot.integrityHash,
  };
}

export function canonicalizeGraphDelta(value: unknown): GraphDelta {
  const delta = validateGraphDelta(value);
  return {
    schema: delta.schema,
    scenarioInstanceId: delta.scenarioInstanceId,
    fromCursor: delta.fromCursor,
    toCursor: delta.toCursor,
    generatedAt: delta.generatedAt,
    upsertNodes: delta.upsertNodes
      .map(canonicalNode)
      .sort((left, right) => compare(left.id, right.id)),
    closeNodeIds: sortedUnique(delta.closeNodeIds),
    upsertEdges: delta.upsertEdges
      .map(canonicalEdge)
      .sort((left, right) => compare(left.id, right.id)),
    closeEdgeIds: sortedUnique(delta.closeEdgeIds),
    evidenceRefs: sortedUnique(delta.evidenceRefs),
    integrityHash: delta.integrityHash,
  };
}

File: services/organization-graph/internal/graphcontract/canonical.go
package graphcontract

import (
	"sort"

	"github.com/Negentropy-Laby/OpenSlack/services/organization-graph/internal/graphjson"
)

func less(left, right string) bool {
	return graphjson.UTF16Less(left, right)
}

func sortedUnique(values []string) []string {
	seen := make(map[string]struct{}, len(values))
	result := make([]string, 0, len(values))
	for _, value := range values {
		if _, exists := seen[value]; exists {
			continue
		}
		seen[value] = struct{}{}
		result = append(result, value)
	}
	sort.Slice(result, func(left, right int) bool { return less(result[left], result[right]) })
	return result
}

func canonicalProperties(value graphjson.Value) graphjson.Value {
	switch current := value.(type) {
	case graphjson.Array:
		result := make(graphjson.Array, len(current))
		for index, item := range current {
			result[index] = canonicalProperties(item)
		}
		return result
	case graphjson.Object:
		result := make(graphjson.Object, len(current))
		for key, item := range current {
			result[key] = canonicalProperties(item)
		}
		return result
	default:
		return current
	}
}

func cloneOptionalString(value *string) *string {
	if value == nil {
		return nil
	}
	cloned := *value
	return &cloned
}

func cloneOptionalAuthority(value *AuthorityRef) *AuthorityRef {
	if value == nil {
		return nil
	}
	cloned := *value
	return &cloned
}

func canonicalOwners(values []ActorRef) ([]ActorRef, error) {
	keyed := make(map[string]ActorRef, len(values))
	for _, value := range values {
		value.DisplayName = cloneOptionalString(value.DisplayName)
		encoded, err := graphjson.Encode(ActorValue(value))
		if err != nil {
			return nil, err
		}
		keyed[string(encoded)] = value
	}
	result := make([]ActorRef, 0, len(keyed))
	for _, value := range keyed {
		result = append(result, value)
	}
	sort.Slice(result, func(left, right int) bool {
		leftName, rightName := "", ""
		if result[left].DisplayName != nil {
			leftName = *result[left].DisplayName
		}
		if result[right].DisplayName != nil {
			rightName = *result[right].DisplayName
		}
		return less(
			result[left].Kind+"\x00"+result[left].ID+"\x00"+leftName,
			result[right].Kind+"\x00"+result[right].ID+"\x00"+rightName,
		)
	})
	return result, nil
}

func canonicalNode(value Node) (Node, error) {
	owners, err := canonicalOwners(value.Owners)
	if err != nil {
		return Node{}, err
	}
	value.Owners = owners
	value.Status = cloneOptionalString(value.Status)
	value.ValidTo = cloneOptionalString(value.ValidTo)
	value.Properties = canonicalProperties(value.Properties).(graphjson.Object)
	value.SourceEventIDs = sortedUnique(value.SourceEventIDs)
	value.EvidenceRefs = sortedUnique(value.EvidenceRefs)
	return value, nil
}

func canonicalEdge(value Edge) Edge {
	value.AuthorityRef = cloneOptionalAuthority(value.AuthorityRef)
	value.ValidTo = cloneOptionalString(value.ValidTo)
	value.SourceEventIDs = sortedUnique(value.SourceEventIDs)
	value.EvidenceRefs = sortedUnique(value.EvidenceRefs)
	return value
}

func CanonicalCompleteness(value Completeness) Completeness {
	value.SourcesRequested = sortedUnique(value.SourcesRequested)
	value.SourcesObserved = sortedUnique(value.SourcesObserved)
	value.MissingSources = sortedUnique(value.MissingSources)
	value.Warnings = sortedUnique(value.Warnings)
	return value
}

func CanonicalizeSnapshot(value Snapshot) (Snapshot, error) {
	if err := ValidateSnapshot(value); err != nil {
		return Snapshot{}, err
	}
	result := value
	result.Nodes = make([]Node, len(value.Nodes))
	for index, node := range value.Nodes {
		canonical, err := canonicalNode(node)
		if err != nil {
			return Snapshot{}, err
		}
		result.Nodes[index] = canonical
	}
	sort.Slice(result.Nodes, func(left, right int) bool { return less(result.Nodes[left].ID, result.Nodes[right].ID) })
	result.Edges = make([]Edge, len(value.Edges))
	for index, edge := range value.Edges {
		result.Edges[index] = canonicalEdge(edge)
	}
	sort.Slice(result.Edges, func(left, right int) bool { return less(result.Edges[left].ID, result.Edges[right].ID) })
	result.Completeness = CanonicalCompleteness(value.Completeness)
	return result, nil
}

func CanonicalizeDelta(value Delta) (Delta, error) {
	if err := ValidateDelta(value); err != nil {
		return Delta{}, err
	}
	result := value
	result.UpsertNodes = make([]Node, len(value.UpsertNodes))
	for index, node := range value.UpsertNodes {
		canonical, err := canonicalNode(node)
		if err != nil {
			return Delta{}, err
		}
		result.UpsertNodes[index] = canonical
	}
	sort.Slice(result.UpsertNodes, func(left, right int) bool {
		return less(result.UpsertNodes[left].ID, result.UpsertNodes[right].ID)
	})
	result.CloseNodeIDs = sortedUnique(value.CloseNodeIDs)
	result.UpsertEdges = make([]Edge, len(value.UpsertEdges))
	for index, edge := range value.UpsertEdges {
		result.UpsertEdges[index] = canonicalEdge(edge)
	}
	sort.Slice(result.UpsertEdges, func(left, right int) bool {
		return less(result.UpsertEdges[left].ID, result.UpsertEdges[right].ID)
	})
	result.CloseEdgeIDs = sortedUnique(value.CloseEdgeIDs)
	result.EvidenceRefs = sortedUnique(value.EvidenceRefs)
	return result, nil
}
